assets/src/components/groupedAutocomplete.tsx

```diff
@@ -0,0 +1,731 @@
+  MoveToEnd = "MoveToEnd",
+  ForceCursorTo = "ForceCursorTo",
+  DeleteCursor = "DeleteCursor",
+  MoveNext = "MoveNext",
+  MovePrev = "MovePrev",
+  FindNearestValidLocation = "FindNearestValidLocation",
+}
+
+/**
+ * {@link cursorLocationReducer} commands and arguments.
+ */
+type CursorStateAction = {
+  lengthsContext: LengthsContext
+  onCursor?: AutocompleteCursorEventProps
+} & (
+  | {
+      action: CursorLocationAction
+    }
+  | {
+      forceLocation: CursorState
+    }
+)
+
+/**
+ * A State Reducer that moves a {@link CursorState cursor} to leaf coordinates
+ * within a 2d jagged array defined by a {@link LengthsContext}.
+ */
+const cursorLocationReducer = (
+  cursorState: CursorState,
+  nextAction: CursorStateAction
+): CursorState => {
+  const { lengthsContext, onCursor } = nextAction
+
+  // Curry `lengthsContext` into `getValidLocationOrUnset`
+  const getValidLocationOrUnsetWithLengthsContext = (
+    proposedLocation: CursorState
+  ) => getValidLocationOrUnset(proposedLocation, lengthsContext)
+
+  // Early exit if we're simply setting the position
+  if ("forceLocation" in nextAction) {
+    return getValidLocationOrUnsetWithLengthsContext(nextAction.forceLocation)
+  }
+
+  const { action } = nextAction
+
+  // Do actions that don't require knowing the type of the `cursorState`
+  switch (action) {
+    case CursorLocationAction.DeleteCursor: {
+      return undefined
+    }
+
+    case CursorLocationAction.MoveToStart: {
+      // In case the length of the arrays are zero
+      // Check if the first value is valid
+      return getValidLocationOrUnsetWithLengthsContext({ group: 0, option: 0 })
+    }
+
+    case CursorLocationAction.MoveToEnd: {
+      // Check if the last value of each length is valid
+      return getValidLocationOrUnsetWithLengthsContext({
+        group: lengthsContext.length - 1,
+        // move out of bounds ( < 0) if it doesn't exist
+        option: lengthsContext[lengthsContext.length - 1] - 1,
+      })
+    }
+  }
+
+  // Do actions that require the `cursorState`
+
+  // Return the current state if it's invalid
+  if (!cursorState) {
+    return cursorState
+  }
+
+  switch (action) {
+    case CursorLocationAction.FindNearestValidLocation: {
+      // Constrain current cursor to bounds of `lengthsContext`
+      return findNearestValidLocation()
+    }
+
+    case CursorLocationAction.MoveNext: {
+      const validCursorLocation =
+        getValidLocationOrUnsetWithLengthsContext({
+          ...cursorState,
+          option: cursorState.option + 1,
+        }) ||
+        // Then the first option in the next group
+        getValidLocationOrUnsetWithLengthsContext({
+          group: cursorState.group + 1,
+          option: 0,
+        })
+
+      if (validCursorLocation !== undefined) {
+        return validCursorLocation
+      }
+
+      // Otherwise return the current state as we've reached the end
+      onCursor?.onCursorExitEdge?.(CursorExitDirection.ExitEnd)
```

Comment:
I couldn't think of a way to to this outside of the reducer, but I'm open to suggestions.

This was for making it possible to refocus the input element after the cursor tries to move past the last or first option.
See https://github.com/mbta/skate/blob/da2eb59366c7669c7cdef03f6c189de76d166725/assets/src/components/searchForm.tsx#L308-L309 .

Reply:
Okay so I tried implementing a way I thought would work, but in the end I couldn't quite figure out how to keep the same functionality.
I've pushed up my WIP at b699eb24a8a4c4874b90fe82e4865bb5e7759f89, and I added a condition to the test to fire two keydown events in a row, instead of just one, which shows the problem I'm facing with implementing this.

Essentially, I need the cursor state to exist for one render, and I'm not sure how to go about that at this moment.
If you have any thoughts on how to approch this or where I have gone wrong I'd appreciate the insight, but right now I think I need to take a break and come back to it, and may need to leave this to be addressed in the future.

Reply:
Hmmm yeah I see what you mean! A simple change doesn't jump out at me, I think this is fine as-is without trying to move the side-effect.